### ISSUE TITLE
Hyperparameter tuning (example with TCNModel) + historical_forecasts() method

### ISSUE THREAD

**Issue by javitausia** [unit8co/darts#1195]
I am trying to get the optimal combination of hyperparameters for my model, but I don't know if I am using the TCNModel().gridsearch() classmethod correctly.
I am first creating the model:
`darts_tcn = TCNModel(
    input_chunk_length = 48,
    output_chunk_length = 24,
    kernel_size = 3,
    num_filters = 15,
    dilation_base = 2,
    dropout = 0.2,
    likelihood = QuantileRegression(quantiles=quantiles),
    torch_metrics = MetricCollection([MeanSquaredError(squared=False),MeanAbsoluteError()]),
    batch_size = 32,
    n_epochs = 15,
    pl_trainer_kwargs    = {"accelerator": "gpu", "devices":[0], "callbacks": [my_stopper]}
)` 
and then I use .gridsearch(), as follows:
`best_tcn_model = darts_tcn.gridsearch(
    parameters=deep_tcn_parameters,
    series=series.split_before(0.7)[0],
    past_covariates=covariates,
    val_series=series.split_before(0.7)[1],
    verbose=True,
    metric=mse,
    n_jobs=-1
)`
where `deep_tcn_parameters`:
`dict(
    input_chunk_length = [48,48*2,48*4],
    output_chunk_length = [24],
    kernel_size = [3,5,7,9],
    num_filters = [15,25,50],
    dilation_base = [2],
    dropout = [0.2],
    likelihood = [QuantileRegression(quantiles=quantiles), GaussianLikelihood()],
    batch_size = [32],
    n_epochs = [10],
    pl_trainer_kwargs = [{"accelerator": "gpu", "devices":[0]}]
)`
and I am getting results, but I don't know if there is a better way of doing it, as I am rewriting some model features, as the fact that it is going to train in the GPU, that I would imagine that is inherent of the model, and I wouldn't have to pass it as a hyperparameter again. It is working tho, but don't know if I am doing it alright.

I also saw in your documentation this example using [RayTune](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html#hyperparameter-tuning-with-ray-tune), which is more tricky and I might give it a try, but I don't know if you recently added more alternatives.

Moreover, is there any way of producing the historical forecasts faster than producing single predictions for each moment in history?? I know this is the way everything is thought, but it takes some time, and maybe there is another option.

Huge congratulations for the library, it is amazing the amount of things you can do, and the amount of time that the user is able to save just using all the things you created! ;)

**Comment by hrzn** [comment]
Thanks for the feedback! A few notes / answers:
* `gridsearch` is a static method so you should call it on the class directly. Something like `best_model, best_params = TCNModel.gridsearch(my_params)`.
* Gridsearch is only providing very basic hyper-parameter search. For anything sophisticated I would recommend relying on other libraries such as Optuna or RayTune. I am currently writing a tutorial/example with Optuna, so if you wait a little you'll be able to read this and get an example.
* AFAIK there's no obvious way to speed up historical forecasts (besides avoiding retraining and using a stride maybe). If you have some suggestions let us know.

**Comment by javitausia** [comment]
Thanks for the quick answer!

Regarding the **hyperparameters optimization**, RayTune [example](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html#hyperparameter-tuning-with-ray-tune) is awesome, but how can I know when the optuna tutorial is ready, so I can have a look at both and see which one fits me better.

For the **historical forecasts**, I know the way all this library is thought relies on predicting series of data, but I was thinking in the way `.predict()` works in libraries such as sklearn or pytorch. I was thinking that maybe I could access the `pl.LightningModule`, and then use `.predict()`, with the TimeSeries from darts, converted into a tensor / datamodule / dataloader that can work with this predict pytorch function, but I can't access the `pl.LightningModule` from the darts model. I am trying to do it with `TCNModel.trainer`, or `TCNModel.trainer.lightning_module`, but soon or after I get this error:
`Unwrapping the module did not yield a `LightningModule`, got <class 'pytorch_lightning.trainer.trainer.Trainer'> instead.`. Maybe all this is useless, but I think it will be interesting to try to implement this way of quickly producing the predicted series, using the pre-built functions of the underlying libraries.

I will keep working on it and see if I can get something, and let you know.

Congrats again for the awesome work, and thanks for the quick answer!! ;)

**Comment by hrzn** [comment]
> Regarding the **hyperparameters optimization**, RayTune [example](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html#hyperparameter-tuning-with-ray-tune) is awesome, but how can I know when the optuna tutorial is ready, so I can have a look at both and see which one fits me better.

You can e.g. watch the repo to be notified once we open a PR for this new tutorial, and when it is merged.



> For the **historical forecasts**, I know the way all this library is thought relies on predicting series of data, but I was thinking in the way `.predict()` works in libraries such as sklearn or pytorch. I was thinking that maybe I could access the `pl.LightningModule`, and then use `.predict()`, with the TimeSeries from darts, converted into a tensor / datamodule / dataloader that can work with this predict pytorch function, but I can't access the `pl.LightningModule` from the darts model. I am trying to do it with `TCNModel.trainer`, or `TCNModel.trainer.lightning_module`, but soon or after I get this error:
> `Unwrapping the module did not yield a `LightningModule`, got <class 'pytorch_lightning.trainer.trainer.Trainer'> instead.`. Maybe all this is useless, but I think it will be interesting to try to implement this way of quickly producing the predicted series, using the pre-built functions of the underlying libraries.

You can try `my_model.model` (where `my_model` is a fitted model), which is of type `pl.LightningModule`. I'm not sure there's a `predict()` function though. Also, Darts is really thought of in order to avoid having to deal with arrays and tensors directly when working with time series. There might in some cases things that could be done more efficiently by accessing the underlying PyTorch module(s), but in principles most Darts functions should already be quite efficient. For instance calling `predict()` on a Darts model will call the underlying `forward()` method of the underlying module, and even batch calls for you in some cases, so things will be in general at least as fast as when naively calling `forward()`. Of course feel free to try it though (and let us know if you get anything interesting working).

**Comment by javitausia** [comment]
Okey perfect, I'll keep having a look at the repo!! thanks for the help ;)
